File: backend/app/schemas/weather.py
from datetime import date

from pydantic import BaseModel, Field


class WeatherResponse(BaseModel):
    latitude: float
    longitude: float
    temperature_c: float
    precipitation_probability: float
    wind_speed_kmh: float
    weather_code: int


class RoadConditionResponse(BaseModel):
    # "good", "caution" or "poor".
    level: str
    label: str


class HourlyForecastResponse(BaseModel):
    # Local time at the place, e.g. "2026-09-25T19:00".
    time: str
    temperature_c: float
    weather_code: int
    precipitation_probability: float


class CurrentConditionsResponse(BaseModel):
    latitude: float
    longitude: float
    temperature_c: float
    weather_code: int
    wind_speed_kmh: float
    visibility_m: float | None
    # "excellent", "good", "moderate", "poor", or None without data.
    visibility: str | None
    temperature_max_c: float
    temperature_min_c: float
    precipitation_probability: float
    road_conditions: RoadConditionResponse
    # The next hours, from the current one (when `hours` is asked for).
    next_hours: list[HourlyForecastResponse] = []


class WeatherRequest(BaseModel):
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)


class DailyWeatherRequest(BaseModel):
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    forecast_date: date


class DailyWeatherResponse(BaseModel):
    forecast_date: date
    latitude: float
    longitude: float
    temperature_max_c: float
    temperature_min_c: float
    precipitation_probability: float
    wind_speed_max_kmh: float
    weather_code: int

File: frontend/src/api/weather.ts
import { apiRequest } from '@/api/client';

export type RoadConditionLevel = 'good' | 'caution' | 'poor';

export type VisibilityRating = 'excellent' | 'good' | 'moderate' | 'poor';

export type HourlyForecast = {
  // Local time at the place, e.g. `2026-09-25T19:00`.
  time: string;
  temperature_c: number;
  weather_code: number;
  precipitation_probability: number;
};

export type CurrentConditions = {
  latitude: number;
  longitude: number;
  temperature_c: number;
  // WMO weather code.
  weather_code: number;
  wind_speed_kmh: number;
  visibility_m: number | null;
  visibility: VisibilityRating | null;
  // Today, in the place's own time zone.
  temperature_max_c: number;
  temperature_min_c: number;
  precipitation_probability: number;
  // Weather-based only (no closures or road works).
  road_conditions: { level: RoadConditionLevel; label: string };
  // From the current hour on, when asked for.
  next_hours: HourlyForecast[];
};

/**
 * Weather right now at a place, with today's range and a road rating.
 */
export function getCurrentConditions(
  latitude: number,
  longitude: number,
  // Also fetch the forecast for this many hours from now.
  hours = 0,
) {
  return apiRequest<CurrentConditions>(
    `/weather/now?latitude=${latitude}&longitude=${longitude}&hours=${hours}`,
  );
}
